// tt-shed/lib/shedRouter.js
var express = require("express");
var router = module.exports = express.Router();

function getCompByReq(req){
    var body = req.body;
    return {
        id : body.id,
        code : body.code,
        name : body.name,
        isBase : body.isBase,
        img : body.img,
        spec : body.spec,
        mates : body.mates,
        unit : body.unit,
        remark : body.remark
    }
}

router.get("/shed/shed_mgr", function(req, res){
    res.render("shed/shed_mgr");
});
router.get("/shed/proj_mgr", function(req, res){
    res.render("shed/proj_mgr");
});

//++++++++++++++++++++++shed 开始+++++++++++++++++++++++++++++++++
//++++++++++++++++++++++shed 结束+++++++++++++++++++++++++++++++++

//++++++++++++++++++++++proj 开始+++++++++++++++++++++++++++++++++
function getProjByReq(req){
    var body = req.body;
    return {
        id : body.id,
        name : body.name,
        state : body.state,
        remark : body.remark
    }
}
router.post("/shed/proj/list", function(req, res, next){
    req.dbClient.pageGridByReq("project", {_orderBy : ["name"]}, req, {}, function(err, data){
        if(err) return next(err);
        res.send(data);
    })
});
router.post("/shed/proj/create", function(req, res, next){
    var proj = getProjByReq(req);
    req.dbClient.insert("project", proj, function(err){
        if(err) return next(err);
        res.success();
    });
});
router.post("/shed/proj/update", function(req, res, next){
    var proj = getProjByReq(req);
    req.dbClient.update("project",proj , {id : proj.id}, function(err){
        if(err) return next(err);
        res.success();
    });
});
router.post("/shed/proj/delete", function(req, res, next){
    var comp = getProjByReq(req);
    req.dbClient.del("project", {id : comp.id}, function(err){
        if(err) return next(err);
        res.success();
    });
});
//++++++++++++++++++++++proj 结束+++++++++++++++++++++++++++++++++

//++++++++++++++++++++++exp 开始+++++++++++++++++++++++++++++++++
//++++++++++++++++++++++exp 结束+++++++++++++++++++++++++++++++++

//++++++++++++++++++++++sche 开始+++++++++++++++++++++++++++++++++
//++++++++++++++++++++++sche 结束+++++++++++++++++++++++++++++++++

//++++++++++++++++++++++comp 开始+++++++++++++++++++++++++++++++++
router.get("/shed/comp_mgr", function(req, res){
    res.render("shed/comp_mgr");
});
//零件列表
router.post("/shed/comp/list", function(req, res, next){
    req.dbClient.pageGridByReq("comp", {_orderBy : ["code", "name"]}, req, {code:"=?"}, function(err, data){
        if(err) return next(err);
        res.send(data);
    })
});
router.post("/shed/comp/create", function(req, res, next){
    var comp = getCompByReq(req);
    req.dbClient.insert("comp", comp, function(err){
        if(err) return next(err);
        res.success();
    });
});
router.post("/shed/comp/update", function(req, res, next){
    var comp = getCompByReq(req);
    req.dbClient.update("comp",comp , {id : comp.id}, function(err){
        if(err) return next(err);
        res.success();
    });
});
router.post("/shed/comp/delete", function(req, res, next){
    var comp = getCompByReq(req);
    req.dbClient.del("comp", {id : comp.id}, function(err){
        if(err) return next(err);
        res.success();
    });
});
//零件列表
router.post("/shed/comp/children", function(req, res, next){
    req.dbClient.pageGridByReq("comp", "id <> ? order by code, name", [req.body.parentId], req, {code:"=?"}, function(err, data){
        if(err) return next(err);
        res.send(data);
    })
});
//零件列表
router.post("/shed/comp/addChild", function(req, res, next){
    req.dbClient.pageGridByReq("comp", {_orderBy : ["code", "name"]}, req, {code:"=?"}, function(err, data){
        if(err) return next(err);
        res.send(data);
    })
});
//零件列表
router.post("/shed/comp/deleteChild", function(req, res, next){
    req.dbClient.pageGridByReq("comp", {_orderBy : ["code", "name"]}, req, {code:"=?"}, function(err, data){
        if(err) return next(err);
        res.send(data);
    })
});
//++++++++++++++++++++++comp 结束+++++++++++++++++++++++++++++++++
