// tt-pvl/lib/router/pvlRouter.js
var express = require("express");
var router = module.exports = express.Router();
router.use(function(req, res, next){
    //TODO 在此进行权限的过滤判断
    var pvlDs = req.pvlDs;
    next();
});

//require("./routeAuto").init(router, "pvl");

router.get("/pvl/pvl_mgr", function(req, res){
    res.render("pvl/pvl_mgr");
});

router.post("/pvl/ctrl/list", function(req, res, next){
    req.dbClient.pageGridByReq("tt_pvl", {_orderBy : ["code"]}, req, {code:"=?"}, function(err, data){
        if(err) return next(err);
        res.send(data);
    })
});
function getPvlByReq(req){
    var body = req.body;
    return {
        code : body.code,
        name : body.name,
        url : body.url,
        type : body.type,
        parent : body.parent
    }
}
router.post("/pvl/ctrl/create", function(req, res, next){
    var pvl = getPvlByReq(req);
    req.dbClient.insert("tt_pvl", pvl, function(err){
        if(err) return next(err);
        res.success();
    });
});
router.post("/pvl/ctrl/update", function(req, res, next){
    var pvl = getPvlByReq(req);
    req.dbClient.update("tt_pvl",pvl , {code : pvl.code}, function(err){
        if(err) return next(err);
        res.success();
    });
});
router.post("/pvl/ctrl/delete", function(req, res, next){
    var pvl = getPvlByReq(req);
    req.dbClient.del("tt_pvl", {code : pvl.code}, function(err){
        if(err) return next(err);
        res.success();
    });
});
